optimum/habana/transformers/models/mixtral/modeling_mixtral.py

```diff
@@ -449,6 +450,172 @@ def gaudi_mixtral_block_sparse_moe_forward(self, hidden_states: torch.Tensor) ->
     return final_hidden_states, router_logits
 
 
+class DynamicFusedMoE(nn.Module):
+    """DynamicFusedMoE layer for MoE models.
+    Based on implementation from vllm:
+    https://github.com/HabanaAI/vllm-fork/blob/84922944da5b65fa43bbdc465650911cbef9de72/vllm/hpu/ops.py#L282"""
+
+    def __init__(self, num_total_experts):
+        super().__init__()
+        self.num_total_experts = num_total_experts
+
+    def forward(
+        self, hidden_states: torch.Tensor, w13: torch.Tensor, w2: torch.Tensor, score: torch.Tensor, topk: int
+    ) -> torch.Tensor:
+        routing_weights, selected_experts = calculate_routing_tensors(score, topk, hidden_states.dtype)
+        # pre-processing for custom op inputs
+        experts_range = range(self.num_total_experts)
+        w13_list = [w13[i, :, :].squeeze() for i in experts_range]
+        w2_list = [w2[i, :, :].squeeze() for i in experts_range]
+        final_hidden_states = torch.ops.hpu.mixture_of_experts(
+            hidden_states=hidden_states,
+            expert_routing_table=selected_experts,
+            router_weights=routing_weights,
+            w12=w13_list,
```

Comment:
do you have a comment on this quote?  https://github.com/huggingface/optimum-habana/pull/1466#issuecomment-2454643810 
"According to the conversation with Ilay, it is not required to manually fuse w1 and w3 into one weight, and **keeping them separately is better than merging them in the case of fine-tuning** considering in 1.19 Synapse release the Custom MoE kernel will support training."